extensions/amp-fx-collection/0.1/providers/parallax.js

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  installPositionObserverServiceForDoc,
+} from '../../../../src/service/position-observer/position-observer-impl';
+import {
+  PositionObserverFidelity,
+} from '../../../../src/service/position-observer/position-observer-worker';
+import {user} from '../../../../src/log';
+import {getServiceForDoc} from '../../../../src/service';
+import {Services} from '../../../../src/services';
+import {setStyles} from '../../../../src/style';
+
+const FACTOR_ATTR = 'data-parallax-factor';
+
+/**
+ * Provides a parallax visual effect given a parallax factor.
+ *
+ * @implements {../amp-fx-collection.FxProviderInterface}
+ */
+export class ParallaxProvider {
+
+  /**
+   * @param  {!../../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    /** @private @const {!../../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc_ = ampdoc;
+
+    /** @private @const {!../../../../src/service/viewport/viewport-impl.Viewport} */
+    this.viewport_ = Services.viewportForDoc(ampdoc);
+
+    installPositionObserverServiceForDoc(ampdoc);
+
+    /** @private @const {!../../../../src/service/position-observer/position-observer-impl.PositionObserver} */
+    this.positionObserver_ = getServiceForDoc(ampdoc, 'position-observer');
+  }
+
+  /**
+   * Installs parallax effect on the element
+   * @param {!Element} element
+   */
+  installOn(element) {
+    Services.vsyncFor(this.ampdoc_.win).measure(() => {
+      new ParallaxElement(element, this.viewport_, this.positionObserver_);
```

Comment:
It's not obvious why this constructor call is wrapped in a `measure`. Is it because the constructor calls `getAdjustedViewportHeight`? If so, I would suggest having `getAdjustedViewportHeight` return a promise from `vsync.measurePromise` instead.

Reply:
done